### PR TITLE
Implement RFC 0008: Language Reference document, packaging, and export (R-34/R-35/R-36)

### PR DESCRIPTION
**Pull request by Evgenus**
Implements [RFC 0008](docs/proposals/0008-language-reference-export.md) — the author-facing Language Reference: document (R-34), packaging (R-35), and the `get_language_reference()` export (R-36). All three ship together with the docstring/README consolidation, per the RFC's atomic-sequencing decision.

## What changed

**R-34 — `docs/LANGUAGE.md` (new).** The template-language reference for authors (human or agent), carrying **cross-cutting semantics only**: templates & the marker, context & scoping (incl. the variable-visibility table), the `NO_CONTENT` model, the error taxonomy, `expr`/`call` machinery, and composition patterns. **No per-entity sections** — per-rule/operator/function prose stays in the registration docs, so the document changes only when the language model changes, not when the catalog grows.

Consolidation (move, don't copy — one owner per fact):

- `SPECIFICATION.md` **§2** reduced to the engine-internal view (walk dispatch, `Context` copy-on-write invariants, `NoContent` implementation, error-plumbing contract); **§4** collapsed to a pointer plus the retained *Recursion budget* invariant; **§11** points at the reference's worked example; **§5** now names `LANGUAGE.md` as the one hand-written artifact and documents the new export as **§5.2**.
- **§4's per-rule facts folded into `transon/rules.py` docstrings** (they grow richer): accessor out-of-scope errors, `attr`'s `NO_CONTENT`-vs-`TransformationError` split, full `map`/`filter` mode semantics, `zip`/`join`/`file`/`format` edge cases, `expr`/`call` empty-`values` errors — plus the one sanctioned docstring→reference pointer for the shared `expr`/`call` application semantics. The `set` docstring trades its deep scoping model (now reference-owned) for a compact per-rule statement.
- **`Transformer` class docstring slimmed ~225 → ~75 lines**: embedder-only (usage, constructor options, error catching, extending). The pitch/install/comparison content is owned solely by `README.md` (already there — no README change needed).

**R-35 — packaging.** `transon/resources/LANGUAGE.md` ships in the wheel and sdist (hatchling picks it up with zero config change — verified by building and inspecting both). `docs/LANGUAGE.md` stays the canonical hand-edited source; identity between the two is test-enforced.

**R-36 — `transon/reference.py`.** `get_language_reference()` → `{reference_version, engine_version, format, content, sections}` with `REFERENCE_VERSION '1.0'`, fence-aware deterministic `##`-heading splitting, GitHub-style slug ids with collision suffixes, the preamble rule, and a `python -m transon.reference` CLI. Engine-global (base `Transformer` only); language facts only.

## Reviewer notes

- **Exported content changes, shapes don't** (called out in CHANGELOG per the RFC's Sequencing): `get_all_docs()['doc']` shrinks to the embedder narrative; per-rule doc text in `get_all_docs()` and `get_editor_metadata()['docs']` grows. Doc text is contractually opaque to the editor; the docs site absorbs this via **D-20** (`transon-org.github.io`, hard dependency of the release).
- **Drift protection** is a pinned section-id list in `tests/test_reference.py` (not catalog-coverage — the RFC deliberately dropped that): editing `LANGUAGE.md`'s section set requires touching the pin and following the `reference_version` policy. Also tested: sections-concatenation parity, packaging parity via `importlib.resources`, canonical↔packaged identity, splitter edge cases (no preamble, slug collisions, fenced `##`, deeper headings).
- **Canonical↔packaged copy**: chose the RFC's committed-copy + identity-test option over build-time force-include — identical behavior in dev/CI/installed layouts; the test fails loudly if the copies diverge.
- Verified: `pytest` 392 passed, 99 % coverage (not reduced); `python -m transon.docs` reports no `TBD`; `scripts/check_roadmap.py` consistent; wheel + sdist built and inspected.
- Bookkeeping: CHANGELOG under `[Unreleased]`; R-34/35/36 → `done` with Shipped notes; RFC 0008 + proposals index → Implemented (unreleased). Cross-repo follow-ups (out of scope here): D-20 docs-site migration, `transon-authoring` reference pin.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added `get_language_reference()` for offline access to the Language Reference (versioned, deterministic section splitting) plus a CLI to output it as JSON.
* **Documentation**
  * Consolidated author-facing language guidance into the exported Language Reference; updated roadmap/spec/RFC 0008 and clarified docs generation/rendering.
* **Bug Fixes**
  * `map` “items” mode now raises a located definition error when `items` doesn’t evaluate to a list.
  * `transform(..., copy_output=True)` preserves `NO_CONTENT` identity, including when nested in copied outputs.
* **Tests**
  * Added/expanded coverage for reference structure, deterministic IDs, packaging parity, and `NO_CONTENT` identity/validation behavior.

### CODE REVIEW

**Comment by Evgenus** [comment]
Addressed the CodeRabbit review in 830acfb. Disposition of all 10 findings:

**Fixed (7):**
- **Tilde fences** (`reference.py`) — proper CommonMark fence tracking: backtick *and* tilde fences; a closer must match the opening character, be at least as long, and carry no trailing content. Regression tests for `~~~` fences and mismatched/longer closers.
- **Whitespace-only preamble** (`reference.py`) — any non-empty prefix is now emitted as the `preamble` section, so sections-concatenation parity holds unconditionally; the test that locked in the old behavior now asserts parity.
- **`map.items` non-list result** (`rules.py`) — now a located `DefinitionError` (matches the existing `expr`/`call` `values` precedent) instead of iterating dict keys/string chars or leaking `TypeError`. Regression test + CHANGELOG entry (behavior change, flagged per project rules).
- **`copy_output` vs `NO_CONTENT`** (`transformers.py`) — the sentinel branch is never deep-copied: identity preserved for `no_content=Transformer.NO_CONTENT`, and a caller-supplied substitute is returned as-is (it cannot alias the input). Identity tests + CHANGELOG entry.
- **"byte-exact" wording** — spec §5.2 and the RFC now say canonical normalized text (UTF-8, LF).
- **Packaging-parity overclaim** — spec + RFC now scope the test to source-tree parity; wheel/sdist inclusion is verified by building the distributions at release.
- **RFC example / lifecycle drift** — export example now shows the real pinned ids (`preamble`, `the-no_content-model`); RFC status no longer claims "in `main`"; proposals index `Shipped` is `—` until a release names the version.

**Declined (3), with reasons:**
- **Two "future date" findings** (RFC status, roadmap acceptance date): 2026-07-18 is the actual decision/implementation date — the review bot's clock appears to be a day behind. No change.
- **Walk-time mutually-exclusive-mode enforcement** (`rule_map`/`rule_expr`/`rule_call`): this exact trade-off is a recorded project decision — Roadmap **R-04** chose option 1 (opt-in `validate()`/`validate=True` catches ambiguous parameter combinations statically) and explicitly noted walk-time enforcement is *breaking* for templates carrying a redundant ignored parameter. The docstrings' "mutually exclusive" statements are the contract `validate()` enforces; changing lazy dispatch would be a silent behavior change contrary to the roadmap decision. No change.

397 tests pass; `python -m transon.docs` reports no TBD; `scripts/check_roadmap.py` consistent.

File: transon/transformers.py [hunk 1]
```diff
@@ -881,7 +728,12 @@ def transform(self, data, no_content=None, *, copy_output: bool = False):
...
-        if result is self.NO_CONTENT and no_content is not self.NO_CONTENT:
+        if result is self.NO_CONTENT:
+            # Never deepcopy this branch: the sentinel is compared by identity
+            # everywhere, and a substitute is caller-owned (it cannot alias the
+            # input, so `copy_output` has nothing to protect).
+            if no_content is self.NO_CONTENT:
+                return result
             return no_content
         if copy_output:
             return copy.deepcopy(result)
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/transformers.py:748]: Fixed in 58397f7 — confirmed real: a literal template list keeps `NO_CONTENT` elements (only container *rules* skip them), so the sentinel can sit inside a deep-copied result. `NoContent` now defines `__copy__`/`__deepcopy__` returning itself, with a nested-identity regression test and an extended changelog note. 398 tests pass.

File: docs/proposals/0008-language-reference-export.md [hunk 2]
```diff
@@ -1,9 +1,13 @@
 # RFC 0008 — Author-facing Language Reference: document, packaging, and export API
 
-- **Status:** Proposed
+- **Status:** Implemented (2026-07-18, in `main` — unreleased; ships in the next release per Sequencing)
 - **Created:** 2026-07-16
-- **Roadmap:** R-34 (Language Reference document), R-35 (package the reference), R-36 (`get_language_reference()` export) — proposed; add rows to `docs/ROADMAP.md` on acceptance. (R-33 is held by [RFC 0007](0007-builtin-function-library.md).)
-- **Type:** New documentation artifact + packaging + a new read-only export API (`get_language_reference()`) — additive; no change to existing template semantics or existing engine APIs
+- **Amended:** 2026-07-18 — consolidation scope extended to the `Transformer` docstring and
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/proposals/0008-language-reference-export.md]: Partially fixed in 830acfb: the status line no longer claims the change is in `main` (it is on this PR branch). The date itself stands — 2026-07-18 is the actual decision/implementation date; the review clock appears to be a day behind.

File: docs/proposals/README.md [hunk 3]
```diff
@@ -43,7 +43,7 @@ is the *authoritative* record of each work item's status via its **R-number**.
 | [0005](0005-example-corpus-normalization.md) | Normalize exports to one flat example corpus | Implemented | R-31 | v0.1.6 | docs site, `transon-blockly` |
 | [0006](0006-transformer-recursion-depth-budget.md) | Bounded per-level recursion budget (self-`include` depth) | Implemented | R-32 | v0.1.7 | `transon-blockly` |
 | [0007](0007-builtin-function-library.md) | Grow the built-in function library | Implemented | R-33 | v0.1.8 | `transon-authoring`, `transon-blockly` |
-| [0008](0008-language-reference-export.md) | Author-facing Language Reference: doc, packaging, export | Proposed | R-34, R-35, R-36 | — | `transon-authoring`, docs site |
+| [0008](0008-language-reference-export.md) | Author-facing Language Reference: doc, packaging, export | Implemented | R-34, R-35, R-36 | unreleased | `transon-authoring`, docs site |
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/proposals/README.md]: Fixed in 830acfb: the Shipped column is now `—` (a release will name the version, flipping it consistently with the lifecycle table).

File: docs/ROADMAP.md [hunk 4]
```diff
@@ -1022,6 +1025,94 @@ failures to `TransformationError`); `split` rule in `transon/rules.py`; total `i
...
+## Theme G — Author-facing Language Reference (RFC 0008)
+
+> Engine-side counterpart of the `transon-authoring` authority-ladder gap (skill contract
+> rung 2): no author-facing, pinnable, offline-servable language document exists. Design and
+> the full ownership principle (structure in the catalog, per-entity behavior in registration
+> docs, cross-cutting semantics in `LANGUAGE.md`) are recorded in
+> [`proposals/0008-language-reference-export.md`](proposals/0008-language-reference-export.md);
+> accepted 2026-07-18. R-34/R-35/R-36 plus the docstring/README consolidation ship
+> **atomically in one release** (RFC Sequencing); the docs-site counterpart is D-20 in
+> [`DOCS_SITE_ROADMAP.md`](DOCS_SITE_ROADMAP.md).
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/ROADMAP.md:1037]: Not changed: 2026-07-18 is the actual acceptance date recorded at decision time; the review clock appears to be a day behind.

File: docs/SPECIFICATION.md [hunk 5]
```diff
@@ -600,6 +405,30 @@ into a lean structural `catalog` (consumed by the editor's generators) and an
...
+`transon/reference.py` provides `get_language_reference()` — a dedicated, versioned
+export (RFC 0008, R-36) serving the packaged `LANGUAGE.md` offline:
+
+- Shape: `{reference_version, engine_version, format: "markdown", content, sections}`.
+  `content` is the byte-exact document (UTF-8, `\n` newlines); `sections` is a flat,
+  ordered split on top-level `##` headings (each section includes its own heading
+  line; deeper headings stay inside their parent; a non-empty intro before the first
+  `##` becomes a leading `{"id": "preamble"}` section; ids are GitHub-style slugs,
+  collisions suffixed `-2`, `-3`, … in document order). Concatenating `sections`
+  reproduces `content` exactly.
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/SPECIFICATION.md]: Fixed in 830acfb: spec §5.2 and the RFC (Deliverable 3) now describe `content` as canonical normalized text (UTF-8, line endings normalized to LF); exactness is stated for the ordered sections-concatenation parity.

File: docs/SPECIFICATION.md [hunk 6]
```diff
@@ -600,6 +405,30 @@ into a lean structural `catalog` (consumed by the editor's generators) and an
...
+- `REFERENCE_VERSION` policy (mirrors `METADATA_VERSION`): additive changes — a new
+  section, appended prose, a new optional field — bump the minor; removing/renaming a
+  section `id`, changing the `sections` shape, or dropping/renaming a top-level field
+  is breaking and bumps the major. Consumers MUST fail loudly on an unsupported major.
+- The export is **engine-global** (base `Transformer` only; no `cls=` parameter) and
+  states language facts only.
+- Packaging: `transon/resources/LANGUAGE.md` ships in the wheel and sdist;
+  `docs/LANGUAGE.md` is the canonical hand-edited source and
+  `tests/test_reference.py` asserts the two are identical, pins the section-id list,
+  and checks the split parity.
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/SPECIFICATION.md]: Fixed in 830acfb: spec §5.2 and the RFC acceptance paragraph now scope the test to **source-tree parity** (packaged resource via `importlib.resources` equals both the export content and the canonical `docs/LANGUAGE.md`); wheel/sdist inclusion is verified by building the distributions at release, not claimed as CI coverage.

File: transon/reference.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    order. The concatenation of all sections reproduces ``content`` exactly.
+    """
+    lines = content.splitlines(keepends=True)
+    boundaries = []
+    in_fence = False
+    for index, line in enumerate(lines):
+        if line.lstrip().startswith('```'):
+            in_fence = not in_fence
+            continue
+        if not in_fence and line.startswith('## '):
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/reference.py]: Fixed in 830acfb: `_split_sections` now tracks the opening fence character and length via a CommonMark-style `_FENCE` matcher — backtick and tilde fences both recognized; a closer must match the opening character, be at least as long, and carry no trailing content. Regression tests added for `~~~` fences and mismatched/longer closers.

File: transon/reference.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+    first = boundaries[0] if boundaries else len(lines)
+    preamble = ''.join(lines[:first])
+    if preamble.strip():
+        sections.append({
+            'id': _unique('preamble'),
+            'title': '',
+            'heading_level': None,
+            'content': preamble,
+        })
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/reference.py:107]: Fixed in 830acfb: the guard is now `if preamble:` so any non-empty prefix — whitespace included — is emitted as the preamble section; the test that locked in the old behavior now expects `[preamble, first]` and asserts concatenation reproduces the input exactly. RFC/spec wording updated to match.

File: transon/rules.py [hunk 9]
```diff
@@ -311,6 +311,20 @@ def _iter_contexts(t: Transformer, context: Context):
...
+
+    - `item`: one output element per input element → list.
+    - `items`: the template yields a *list* of elements per input element; the
+      lists are concatenated → list (the flatten-while-mapping mode).
+    - `key`+`value`: one output entry per input element → dict.
+
+    Results that evaluate to `NO_CONTENT` are skipped — the item (or the whole
+    key/value pair) is omitted from the output rather than emitted as `null`.
+    Iterating a value that is neither a list nor a dict raises
+    `TransformationError`.
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/rules.py:327]: First part fixed in 830acfb: `map.items` now validates the evaluated value is a list and raises a located `DefinitionError` (matching the existing `expr`/`call` `values` precedent) instead of iterating dict keys/string chars or leaking a raw `TypeError`; regression test + changelog entry added. Second part (walk-time mutual-exclusion enforcement) declined: Roadmap R-04 explicitly chose opt-in `validate()` over walk-time enforcement because the latter is breaking for templates carrying a redundant ignored parameter — the docstrings state the contract that `validate()` enforces.

File: transon/transformers.py [hunk 10]
```diff
@@ -292,165 +267,37 @@ class Transformer:
...
+      (see :class:`IncludeContext`).
+    - **`max_include_depth`** — nested-`include` depth limit (default 50).
+
+    `transform(data, no_content=None, *, copy_output=False)` returns the output:
+    `no_content` chooses what a top-level `NO_CONTENT` becomes (default `None`);
+    `copy_output=True` deep-copies the result once so it shares no mutable
+    structure with the input (which is never mutated regardless — but pass-through
+    rules return references into it). Failures raise `DefinitionError` (malformed
+    template) or `TransformationError` (data does not fit) — catch these two;
+    messages include the template path (`at template → …`).
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply transon/transformers.py:296]: Fixed in 830acfb (top-level identity: the sentinel branch returns before `deepcopy`, and a caller-owned `no_content` substitute is returned as-is) and completed in 58397f7 (nested identity: `NoContent.__copy__`/`__deepcopy__` return `self`). Identity regression tests for both cases; changelog entry added.

File: docs/proposals/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/proposals/README.md]: Fixed in c728598 by amending the lifecycle definition rather than the entry: the roadmap-term mapping in the same table requires Implemented ↔ `done`, and R-34/35/36 are `done` (implementation merged), so flipping the RFC back to Accepted would create the inverse inconsistency. **Implemented** now means *implementation merged (roadmap items `done`); the `Shipped` column names the release version once tagged — `—` until then*. The 0008 row (`Implemented` / `—`) is consistent with that definition, and the version lands in `Shipped` at release.

File: docs/SPECIFICATION.md [hunk 11]
```diff
@@ -43,6 +46,14 @@ Packaging is uv / PEP 621 (`pyproject.toml`, `uv.lock`). Runtime dependencies: n
...
 ## 2. Core concepts
 
+> **Deliberate duplication (decision 2026-07-18).** This section is the complete
+> normative statement of the language semantics *inside the engine contract*; the
+> author-facing packaged form is [`transon/resources/LANGUAGE.md`](../transon/resources/LANGUAGE.md)
+> (served by `transon.reference.get_language_reference()`, §5.2), and per-entity
+> behavior is also stated in the registration docs. The spec deliberately keeps its
+> own full copy so it remains a single, complete document — when changing behavior,
+> update all three surfaces in the same change.
+
```
[the remark this answers is not in the document]
Comment (Evgenus) [reply docs/SPECIFICATION.md:56]: Not changed — same resolution as the two sibling date findings you withdrew earlier in this review (on the RFC status line and the roadmap acceptance date): 2026-07-18 is the actual decision date; the review clock appears to be a day behind. The spec banners cite the same decision, so the same date is correct.
